dropshot/src/error.rs

```diff
@@ -50,6 +50,89 @@ use serde::Serialize;
 use std::error::Error;
 use std::fmt;
 
+pub use crate::extractor::ExtractorError;
+pub use crate::extractor::MultipartBodyError;
+pub use crate::extractor::QueryError;
+pub use crate::extractor::StreamingBodyError;
+pub use crate::extractor::TypedBodyError;
+pub use crate::http_util::PathError;
+pub use crate::router::RouterError;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ServerError {
+    #[error(transparent)]
+    Route(#[from] RouterError),
+    #[error(transparent)]
+    Extractor(#[from] crate::error::ExtractorError),
+    #[error("response error: {0}")]
+    Response(ResponseError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ResponseError {
+    #[error(transparent)]
+    Serialize(#[from] serde_json::Error),
+    #[error(transparent)]
+    Body(#[from] http::Error),
+    #[error("error processing headers: {0}")]
+    HeaderMap(String),
+    #[error(transparent)]
+    HeaderName(#[from] http::header::InvalidHeaderName),
+    #[error("invalid value for {} header", .name)]
+    HeaderValue {
+        name: http::header::HeaderName,
+        // This is skipped from the display output, because it just says
+        // "invalid header value" again.
+        #[source]
+        error: http::header::InvalidHeaderValue,
+    },
+}
+
+impl From<ServerError> for HttpError {
+    fn from(error: ServerError) -> Self {
+        match error {
+            ServerError::Route(e) => e.into(),
+            ServerError::Extractor(e) => e.into(),
+            ServerError::Response(e) => e.into(),
+        }
+    }
+}
+
+impl From<ResponseError> for HttpError {
+    fn from(error: ResponseError) -> Self {
+        HttpError::for_internal_error(error.to_string())
+    }
+}
+
+/// Trait implemented by errors which can be converted into an HTTP response.
+///
+/// In order to be returned as an error from a Dropshot endpoint handler, a type
+/// must implement `AsStatusCode` (along with [`Serialize`] and [`JsonSchema`]).
+/// This trait is implemented by [`HttpError`], and may alternatively be
+/// implemented by user-defined error types.
+pub trait IntoErrorResponse: JsonSchema {
+    /// Returns the HTTP status code that should be returned along with this
+    /// error.
+    fn into_error_response(
+        &self,
+        request_id: &str,
+    ) -> http::Response<crate::Body>;
+}
```

Comment:
This is the trait that must be implemented by error types returned by endpoint functions. Initially, they were just going to be `T: Serialize + JsonSchema + AsStatusCode` (where `AsStatusCode` is a trait where the user can indicate the desired status code for the error, naturally), but I realized that we might also want to provide the request ID when turning user error types into error responses.

I'm wondering about unifying this interface with that of the server's global handler for dropshot-internal errors, but it's tricky: we need to erase the types of the errors returned by individual endpoint functions, in order to allow them to return different error types without forcing the entire server to be generic over every possible error, but we'd like the global error handler function to be able to match on the dropshot-internal errors. I thought about having this function take take the `ErrorContext` type used by the global error handler thingy, but that's currently generic over the `ServerContext`, and would therefore make this trait no longer object-safe, defeating the purpose a bit.